### PR TITLE
Improve ENF isConditional

### PR DESCRIPTION
**Pull request by gnawf**

### CODE REVIEW

File: src/main/java/graphql/normalized/ExecutableNormalizedField.java [hunk 1]
```diff
@@ -505,14 +502,69 @@ private void traverseImpl(ExecutableNormalizedField root,
...
 
+    /**
+     * This tries to find interfaces common to all the field output types.
+     * <p>
+     * i.e. goes through {@link #getFieldDefinitions(GraphQLSchema)} and finds an interface that
+     * all the field's unwrapped output types are assignable to.
+     */
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    private Set<GraphQLInterfaceType> getInterfacesCommonToAllOutputTypes(GraphQLSchema schema) {
+        Ref<Set<GraphQLInterfaceType>> ref = new Ref<>();
```
Comment (bbakerman) [inline src/main/java/graphql/normalized/ExecutableNormalizedField.java:513]: Why do you need this Ref class?

Set<GraphQLInterfaceType>> set = null;

if (set = null) {
  set = new LinkedHashSet()
} else {
  set.retainAll(x)
}

File: src/main/java/graphql/util/Ref.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package graphql.util;
+
+import graphql.Internal;
+
+@Internal
+public class Ref<T> {
+    public T value;
+}
```
Comment (bbakerman) [inline src/main/java/graphql/util/Ref.java:8]: I dont see how this is useful?

File: src/main/java/graphql/normalized/ExecutableNormalizedField.java (hunk 1 above)
Comment (gnawf) [reply src/main/java/graphql/normalized/ExecutableNormalizedField.java:513]: It's Java. Can't touch non (effectively) final variables from inside lambdas.

(This gets used inside the `forEachFieldDefinitions` lambda).

It's basically a replacement for `AtomicReference` or `Set<GraphQLInterfaceType>[]` but more performant.

I will however move this down closer to the lambda usage.

**Comment by gnawf** [comment]
Benchmarks using a large query.

Old code

```
Benchmark                                            Mode  Cnt    Score   Error  Units
AstCompilerBenchmark.benchMarkAstPrinterThroughput  thrpt   15  253.471 ± 4.965  ops/s
```

New code

```
Benchmark                                            Mode  Cnt    Score    Error  Units
AstCompilerBenchmark.benchMarkAstPrinterThroughput  thrpt   15  414.082 ± 46.620  ops/s
```

File: src/main/java/graphql/util/Ref.java (hunk 2 above)
Comment (bbakerman) [reply src/main/java/graphql/util/Ref.java:8]: I get it now

File: src/main/java/graphql/normalized/ExecutableNormalizedField.java (hunk 1 above)
Comment (bbakerman) [reply src/main/java/graphql/normalized/ExecutableNormalizedField.java]: Ahh I get it - another trick is an array with one entry

```
Set<GraphQLInterfaceType> ref[] = {null}
```
and then use the one entry in a mutable way.

I like your approach.

File: src/main/java/graphql/util/Ref.java (hunk 2 above)
Comment (bbakerman) [reply src/main/java/graphql/util/Ref.java:8]: Maybe if this was named `MutableRef` it might be clearer on how it can be used

**Review by bbakerman — APPROVED** [review]
Small naming suggestion on Ref - I like how this changed no tests and it still works

I am pretty sure there is lots of conditional tests

**Pull request merged**